sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserver.java

```
@@ -0,0 +1,109 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.vendor.grpc.v1p21p0.io.grpc.stub.StreamObserver;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * A buffering outbound {@link FnDataReceiver} with both size-based buffer and time-based buffer
+ * enabled for the Beam Fn Data API.
+ */
+public class BeamFnDataTimeBasedBufferingOutboundObserver<T>
+    extends BeamFnDataSizeBasedBufferingOutboundObserver<T> {
+
+  private final Object lock;
+  private final ScheduledFuture<?> flushFuture;
+  @VisibleForTesting final AtomicReference<IOException> flushException;
+
+  BeamFnDataTimeBasedBufferingOutboundObserver(
+      int sizeLimit,
+      long timeLimit,
+      LogicalEndpoint outputLocation,
+      Coder<T> coder,
+      StreamObserver<BeamFnApi.Elements> outboundObserver) {
+    super(sizeLimit, outputLocation, coder, outboundObserver);
+    this.lock = new Object();
+    this.flushFuture =
+        Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                    .setDaemon(true)
+                    .setNameFormat("DataBufferOutboundFlusher-thread")
+                    .build())
+            .scheduleAtFixedRate(this::periodicFlush, timeLimit, timeLimit, TimeUnit.MILLISECONDS);
+    this.flushException = new AtomicReference<>(null);
+  }
+
+  @Override
+  public void close() throws Exception {
+    checkFlushThreadException();
+    synchronized (lock) {
+      flushFuture.cancel(true);
+      try {
+        flushFuture.get();
+      } catch (CancellationException | ExecutionException | InterruptedException exn) {
+        // expected
```

CancellationException is the only exception that is expected. ExecutionException should be unwrapped and we should throw the cause (if set) otherwise we should throw the execution exception directly. InterruptedException should not be caught here at all.